pow/web3j/src/main/java/org/ethereum/beacon/pow/Web3JDepositContract.java

```diff
@@ -0,0 +1,312 @@
+  private static final Logger logger = LogManager.getLogger(Web3JDepositContract.class);
+
+  private final LatestExecutor<Long> blockExecutor;
+  private final Web3j web3j;
+  private final Web3RequestExecutor web3RequestExecutor;
+  private final org.ethereum.beacon.pow.contract.DepositContract depositContract;
+  private final Address contractDeployAddress;
+  private final Hash32 contractDeployAddressHash;
+  private final long contractDeployBlock;
+  private final BloomFilter contractAddressBloom;
+  private volatile long processedUpToBlock;
+  private boolean chainStartComplete;
+
+  public Web3JDepositContract(
+      Web3j web3j,
+      String contractDeployAddress,
+      long contractDeployBlock,
+      Schedulers schedulers,
+      Function<BytesValue, Hash32> hashFunction,
+      int merkleTreeDepth,
+      BeaconChainSpec spec) {
+    super(schedulers, hashFunction, merkleTreeDepth, spec);
+    this.web3j = web3j;
+
+    this.contractDeployAddress = Address.fromHexString(contractDeployAddress);
+    contractDeployAddressHash =
+        Hash32.wrap(Bytes32.wrap(Hash.sha3(this.contractDeployAddress.extractArray())));
+    this.contractAddressBloom = BloomFilter.create(contractDeployAddressHash.extractArray());
+    this.contractDeployBlock = contractDeployBlock;
+    this.depositContract =
+        org.ethereum.beacon.pow.contract.DepositContract.load(
+            contractDeployAddress,
+            web3j,
+            Credentials.create(ECKeyPair.create(BigInteger.ONE)),
+            new DefaultGasProvider());
+
+    processedUpToBlock = contractDeployBlock;
+    blockExecutor = new LatestExecutor<>(this.schedulers.blocking(), this::processBlocksUpTo);
+    this.web3RequestExecutor = new Web3RequestExecutor(web3j);
+  }
+
+  @Override
+  protected void chainStartSubscribed() {
+    web3RequestExecutor.executeOnSyncDone(
+        () -> {
+          processConfirmedBlocks();
+          onEthereumUpdated();
+        });
+    web3j
+        .blockFlowable(false)
+        .doOnNext(
+            block -> {
+              onEthereumUpdated();
+            })
+        .subscribe();
+  }
+
+  @Override
+  protected void chainStartDone() {
+    chainStartComplete = true;
+  }
+
+  private void onEthereumUpdated() {
+    processConfirmedBlocks();
+  }
+
+  long getBestConfirmedBlock() {
+    return web3j.ethBlockNumber().sendAsync().join().getBlockNumber().longValue()
+        - getDistanceFromHead();
+  }
+
+  void processConfirmedBlocks() {
+    long bestConfirmedBlock = getBestConfirmedBlock();
+    blockExecutor.newEvent(bestConfirmedBlock);
+  }
+
+  void processBlocksUpTo(long bestConfirmedBlock) {
+    for (long number = processedUpToBlock; number <= bestConfirmedBlock; number++) {
+      EthBlock block =
+          web3j
+              .ethGetBlockByNumber(DefaultBlockParameter.valueOf(BigInteger.valueOf(number)), false)
+              .sendAsync()
+              .join();
```

Comment:
I am thinking of processing blocks and transaction receipts in an asynchronous fashion. We probably can get a block flow out of web3j, then add a processor that filters out blocks with usage of bloom filter, then yet another one which requests for tx receipts of valuable blocks. I guess we can created a stream upon a set of futures that are requesting for receipts. And at the end parse deposit data from tx receipts upon their arrival. 

This scheme has at least one flaw, deposit events might be processed in the wrong order but I don't think that there will be a big mess. So we could probably have a buffer that could be drained when there is a deposit standing next to last processed one.